Python/fileutils.c

```diff
@@ -1097,13 +1114,56 @@ _Py_attribute_data_to_stat(BY_HANDLE_FILE_INFORMATION *info, ULONG reparse_tag,
     result->st_reparse_tag = reparse_tag;
     if (info->dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
         reparse_tag == IO_REPARSE_TAG_SYMLINK) {
-        /* first clear the S_IFMT bits */
-        result->st_mode ^= (result->st_mode & S_IFMT);
-        /* now set the bits that make this a symlink */
-        result->st_mode |= S_IFLNK;
+        /* set the bits that make this a symlink */
+        result->st_mode = (result->st_mode & ~S_IFMT) | S_IFLNK;
     }
     result->st_file_attributes = info->dwFileAttributes;
 }
+
+void
+_Py_stat_basic_info_to_stat(FILE_STAT_BASIC_INFORMATION *info,
+                            struct _Py_stat_struct *result)
+{
+    /* set values here to match WIN32_STATX_BASIC_STATS defined in posixmodule.c */
+    memset(result, 0, sizeof(*result));
+    result->st_mode = attributes_to_mode(info->FileAttributes);
+    result->st_size = info->EndOfFile.QuadPart;
+    LARGE_INTEGER_to_time_t_nsec(&info->CreationTime, &result->st_btime, &result->st_btime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&info->ChangeTime, &result->st_ctime, &result->st_ctime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&info->LastWriteTime, &result->st_mtime, &result->st_mtime_nsec);
+    LARGE_INTEGER_to_time_t_nsec(&info->LastAccessTime, &result->st_atime, &result->st_atime_nsec);
+    result->st_nlink = info->NumberOfLinks;
+    result->st_ino = info->FileId.QuadPart;
+    /* bpo-37834: Only actual symlinks set the S_IFLNK flag. But lstat() will
+       open other name surrogate reparse points without traversing them. To
+       detect/handle these, check st_file_attributes and st_reparse_tag. */
+    result->st_reparse_tag = info->ReparseTag;
+    if (info->FileAttributes & FILE_ATTRIBUTE_REPARSE_POINT &&
+        info->ReparseTag == IO_REPARSE_TAG_SYMLINK) {
+        /* set the bits that make this a symlink */
+        result->st_mode = (result->st_mode & ~S_IFMT) | S_IFLNK;
+    }
+    result->st_file_attributes = info->FileAttributes;
+    switch (info->DeviceType) {
```

Comment:
TODO: Verify that this is `FILE_TYPE_*` values and not `FILE_DEVICE_*` values.

Reply:
Definitely `FILE_DEVICE_*` values

Reply:
Are there plans for the named-pipe filesystem ("\\\\?\\PIPE\\") to implement the stat info classes? Currently, it doesn't support `FileStatInformation` with `NtQueryInformationFile()` or `NtQueryInformationByName()`. The latter also has the same problem that we currently have for a named pipe with `os.stat()`. It opens the client end of the named pipe, which consumes an instance of the pipe. The number of instances may be limited to 1-254. Subsequently the server has to call `DisconnectNamedPipe()` and `ConnectNamedPipe()` to listen for a new connection on the instance.

The stat info classes won't be supported by non-filesystem devices such as NUL, CON, and COM1, which usually map to `FILE_TYPE_CHAR`. We'll be taking the slow path if it isn't a filesystem.